tap_mailchimp/streams.py

```diff
@@ -0,0 +1,501 @@
+                LOGGER.info('reports_email_activity - Previous run from state expired: %s',
+                            batch_id)
+                return
+
+            # Resume from bookmarked job_id, then if completed, issue a new batch for processing.
+            campaigns = [] # Don't need a list of campaigns if resuming
+            self.sync_email_activities(campaigns, batch_id)
+
+    def poll_email_activity(self, batch_id):
+        """Get the email activity URL after the batch as executed"""
+        sleep = 0
+        start_time = time.time()
+        while True:
+            data = self.get_batch_info(batch_id)
+
+            ## needs to update frequently for target-stitch to capture state
+            self.write_activity_batch_bookmark(batch_id)
+
+            progress = ''
+            if data['total_operations'] > 0:
+                progress = ' ({}/{} {:.2f}%)'.format(
+                    data['finished_operations'],
+                    data['total_operations'],
+                    (data['finished_operations'] / data['total_operations']) * 100.0)
+
+            LOGGER.info('reports_email_activity - Job polling: %s - %s%s',
+                        data['id'],
+                        data['status'],
+                        progress)
+
+            if data['status'] == 'finished':
+                return data
+            elif (time.time() - start_time) > MAX_RETRY_ELAPSED_TIME:
+                message = 'Mailchimp campaigns export is still in progress after {} seconds. Will continue with this export on the next sync.'.format(MAX_RETRY_ELAPSED_TIME)
+                LOGGER.error(message)
+                raise Exception(message)
+
+            sleep = next_sleep_interval(sleep)
+            LOGGER.info('campaigns - status: %s, sleeping for %s seconds',
+                        data['status'],
+                        sleep)
+            time.sleep(sleep)
+
+    def stream_email_activity(self, archive_url):
+        """Get the Email Activities from the provided URL after batch completion"""
+
+        failed_campaign_ids = []
+
+        with self.client.request('GET', url=archive_url, s3=True, endpoint='s3') as response:
+            with tarfile.open(mode='r|gz', fileobj=response.raw) as tar:
+                file = tar.next()
+                while file:
+                    if file.isfile():
+                        raw_operations = tar.extractfile(file)
+                        operations = json.loads(raw_operations.read().decode('utf-8'))
+                        for i, operation in enumerate(operations):
+                            campaign_id = operation['operation_id']
+                            last_bookmark = self.state.get('bookmarks', {}).get(self.stream_name, {}).get(campaign_id)
+                            LOGGER.info("reports_email_activity - [batch operation %s] Processing records for campaign %s", i, campaign_id)
+                            if operation['status_code'] != 200:
```

Comment:
Add code comment to explain this scenario as we are not raising an error for status_code!=200

Reply:
Added comment.